src/clonemap.rs

```diff
@@ -0,0 +1,859 @@
+            assert!(val.is_new());
+        }
+
+        for i in 0..len {
+            let val = map.get_or_insert(i, 0);
+            assert_eq!(i, *val.key());
+            assert_eq!(i, *val.value());
+            assert!(!val.is_new());
+        }
+    }
+
+    #[test]
+    fn get_or_insert_many() {
+        get_or_insert_many_inner(CloneConMap::new(), TEST_BATCH);
+    }
+
+    #[test]
+    fn get_or_insert_collision() {
+        get_or_insert_many_inner(CloneConMap::with_hasher(NoHasher), TEST_BATCH_SMALL);
+    }
+
+    #[test]
+    fn simple_remove() {
+        let map = CloneConMap::new();
+        assert!(map.remove(&42).is_none());
+        assert!(map.insert(42, "hello").is_none());
+        assert_eq!("hello", *map.get(&42).unwrap().value());
+        assert_eq!("hello", *map.remove(&42).unwrap().value());
+        assert!(map.get(&42).is_none());
+        assert!(map.is_empty());
+        assert!(map.remove(&42).is_none());
+        assert!(map.is_empty());
+    }
+
+    fn remove_many_inner<H: BuildHasher>(mut map: CloneConMap<usize, usize, H>, len: usize) {
+        for i in 0..len {
+            assert!(map.insert(i, i).is_none());
+        }
+        for i in 0..len {
+            assert_eq!(i, *map.get(&i).unwrap().value());
+            assert_eq!(i, *map.remove(&i).unwrap().value());
+            assert!(map.get(&i).is_none());
+            map.raw.assert_pruned();
+        }
+
+        assert!(map.is_empty());
+    }
+
+    #[test]
+    fn remove_many() {
+        remove_many_inner(CloneConMap::new(), TEST_BATCH);
+    }
+
+    #[test]
+    fn remove_many_collision() {
+        remove_many_inner(CloneConMap::with_hasher(NoHasher), TEST_BATCH_SMALL);
+    }
+
+    #[test]
+    fn collision_remove_one_left() {
+        let mut map = CloneConMap::with_hasher(NoHasher);
+        map.insert(1, 1);
+        map.insert(2, 2);
+
+        map.raw.assert_pruned();
+
+        assert!(map.remove(&2).is_some());
+        map.raw.assert_pruned();
+
+        assert!(map.remove(&1).is_some());
+
+        map.raw.assert_pruned();
+        assert!(map.is_empty());
+    }
+
+    #[test]
+    fn remove_par() {
+        let mut map = CloneConMap::new();
+        for i in 0..TEST_THREADS * TEST_BATCH {
+            map.insert(i, i);
+        }
+
+        thread::scope(|s| {
+            for t in 0..TEST_THREADS {
+                let map = &map;
+                s.spawn(move |_| {
+                    for i in 0..TEST_BATCH {
+                        let num = t * TEST_BATCH + i;
+                        let val = map.remove(&num).unwrap();
+                        assert_eq!(num, *val.value());
+                        assert_eq!(num, *val.key());
+                    }
+                });
+            }
+        })
+        .unwrap();
+
+        map.raw.assert_pruned();
+        assert!(map.is_empty());
+    }
+
+    #[test]
+    fn unsized_values_require_boxes() {
+        let map: CloneConMap<usize, Box<[usize]>> = CloneConMap::new();
```

Comment:
This is not really unsized. `[usize]` is unsized, but `Box` already has know size. So this doesn't really test anything new.